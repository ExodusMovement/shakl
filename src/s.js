import React from 'react';

import flatten from './flatten';

const s = (Comp, { name, multi, ...opts } = {}) => style => {
  const S = React.forwardRef(({ children, ...props }, ref) => {
    const { attrs = {}, comp, child } = opts;

    const styles = multi ? style : { style };

    return React.createElement(
      comp || Comp,
      {
        ref,
        ...attrs,
        ...props,
        ...Object.keys(styles).reduce(
          (obj, prop) => ({
            ...obj,
            [prop]: flatten([
              attrs[prop],
              typeof styles[prop] === 'function'
                ? styles[prop](props)
                : styles[prop],
              props[prop]
            ])
          }),
          {}
        )
      },
      child ? React.createElement(child, {}, children) : children
    );
  });

  S.displayName = name || `styled(${Comp.displayName || Comp.name})`;

  S.extend = more => s(S, { name })(more);
  S.withProps = attrs => s(S, { name, attrs })();
  S.withComponent = comp => s(S, { name, comp })(style);
  S.withChild = child => s(S, { name, child })();

  return S;
};

export default s;
