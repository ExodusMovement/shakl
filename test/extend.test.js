import React from 'react';

import { create as r } from 'react-test-renderer';

import s from '../src';

it('extends a styled component with extend()', () => {
  const Foo = s.Text({ fontSize: 20 });
  const Bar = Foo.extend({ fontWeight: 'bold' });
  const Baz = Bar.extend({ color: 'red' });

  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();
  const baz = r(<Baz />).toJSON();

  expect(foo.props.style).toEqual({ fontSize: 20 });
  expect(bar.props.style).toEqual({ fontSize: 20, fontWeight: 'bold' });
  expect(baz.props.style).toEqual({
    fontSize: 20,
    fontWeight: 'bold',
    color: 'red'
  });

  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
  expect(baz).toMatchSnapshot();
});

it('extends a styled component with styled(Styled)', () => {
  const Foo = s.Text({ fontSize: 20 });
  const Bar = s(Foo)({ fontWeight: 'bold' });
  const Baz = s(Bar)({ color: 'red' });

  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();
  const baz = r(<Baz />).toJSON();

  expect(foo.props.style).toEqual({ fontSize: 20 });
  expect(bar.props.style).toEqual({ fontSize: 20, fontWeight: 'bold' });
  expect(baz.props.style).toEqual({
    fontSize: 20,
    fontWeight: 'bold',
    color: 'red'
  });

  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
  expect(baz).toMatchSnapshot();
});

it('extends a styled component with null static styles', () => {
  const Foo = s.View();
  const Bar = Foo.extend();

  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();

  expect(foo.props.style).toEqual({});
  expect(bar.props.style).toEqual({});

  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
});

it('extends a styled component with null dynamic styles', () => {
  const Foo = s.View();
  const Bar = Foo.extend(() => null);

  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();

  expect(foo.props.style).toEqual({});
  expect(bar.props.style).toEqual({});

  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
});

it('extends a styled component with empty static styles', () => {
  const Foo = s.View();
  const Bar = Foo.extend({});

  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();

  expect(foo.props.style).toEqual({});
  expect(bar.props.style).toEqual({});

  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
});

it('extends a styled component with empty dynamic styles', () => {
  const Foo = s.View();
  const Bar = Foo.extend(() => ({}));

  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();

  expect(foo.props.style).toEqual({});
  expect(bar.props.style).toEqual({});

  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
});

it('extends a styled component with dynamic styles based on props', () => {
  const Foo = s.Text();
  const Bar = Foo.extend(p => ({ padding: p.padded ? 10 : 0 }));

  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();
  const barPadded = r(<Bar padded />).toJSON();

  expect(bar.props.style).toEqual({ padding: 0 });
  expect(barPadded.props.style).toEqual({ padding: 10 });

  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
  expect(barPadded).toMatchSnapshot();
});

it('extends a styled component with custom props', () => {
  const Foo = s.Text({ fontSize: 20 });
  const Bar = Foo.extend({ fontWeight: 'bold' }).attrs({ bar: true });
  const Baz = Bar.extend({ color: 'red' }).attrs({ baz: true });

  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();
  const baz = r(<Baz />).toJSON();

  expect(bar.props.bar).toBe(true);
  expect(baz.props.bar).toBe(true);
  expect(baz.props.baz).toBe(true);
  expect(foo.props.style).toEqual({ fontSize: 20 });
  expect(bar.props.style).toEqual({ fontSize: 20, fontWeight: 'bold' });
  expect(baz.props.style).toEqual({
    fontSize: 20,
    fontWeight: 'bold',
    color: 'red'
  });

  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
  expect(baz).toMatchSnapshot();
});
