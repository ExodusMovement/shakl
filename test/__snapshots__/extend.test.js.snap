// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extends a styled component with custom props 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "fontSize": 20,
    }
  }
/>
`;

exports[`extends a styled component with custom props 2`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  bar={true}
  ellipsizeMode="tail"
  style={
    Object {
      "fontSize": 20,
      "fontWeight": "bold",
    }
  }
/>
`;

exports[`extends a styled component with custom props 3`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  bar={true}
  baz={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "red",
      "fontSize": 20,
      "fontWeight": "bold",
    }
  }
/>
`;

exports[`extends a styled component with dynamic styles based on props 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={Object {}}
/>
`;

exports[`extends a styled component with dynamic styles based on props 2`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "padding": 0,
    }
  }
/>
`;

exports[`extends a styled component with dynamic styles based on props 3`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  padded={true}
  style={
    Object {
      "padding": 10,
    }
  }
/>
`;

exports[`extends a styled component with empty dynamic styles 1`] = `
<View
  style={Object {}}
/>
`;

exports[`extends a styled component with empty dynamic styles 2`] = `
<View
  style={Object {}}
/>
`;

exports[`extends a styled component with empty static styles 1`] = `
<View
  style={Object {}}
/>
`;

exports[`extends a styled component with empty static styles 2`] = `
<View
  style={Object {}}
/>
`;

exports[`extends a styled component with extend() 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "fontSize": 20,
    }
  }
/>
`;

exports[`extends a styled component with extend() 2`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "fontSize": 20,
      "fontWeight": "bold",
    }
  }
/>
`;

exports[`extends a styled component with extend() 3`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "red",
      "fontSize": 20,
      "fontWeight": "bold",
    }
  }
/>
`;

exports[`extends a styled component with null dynamic styles 1`] = `
<View
  style={Object {}}
/>
`;

exports[`extends a styled component with null dynamic styles 2`] = `
<View
  style={Object {}}
/>
`;

exports[`extends a styled component with null static styles 1`] = `
<View
  style={Object {}}
/>
`;

exports[`extends a styled component with null static styles 2`] = `
<View
  style={Object {}}
/>
`;

exports[`extends a styled component with styled(Styled) 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "fontSize": 20,
    }
  }
/>
`;

exports[`extends a styled component with styled(Styled) 2`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "fontSize": 20,
      "fontWeight": "bold",
    }
  }
/>
`;

exports[`extends a styled component with styled(Styled) 3`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "red",
      "fontSize": 20,
      "fontWeight": "bold",
    }
  }
/>
`;
